swarm/pss/notify/notify.go

```diff
@@ -0,0 +1,276 @@
+		}
+		err = c.pss.SendSym(m.symKeyId, c.notifiers[name].topic, smsg)
+		if err != nil {
+			log.Warn("Failed to send notify to addr %x: %v", m.address, err)
+		}
+	}
+	return nil
+}
+
+// adds an client address to the corresponding address bin in the notifier service
+// this method is not concurrency safe
+func (c *Controller) addToNotifier(name string, address pss.PssAddress) (string, error) {
+	notifier, ok := c.notifiers[name]
+	if !ok {
+		return "", fmt.Errorf("Unknown notifier %s", name)
+	}
+	for _, m := range notifier.bins {
+		if bytes.Equal(address, m.address) {
+			m.count++
+			return m.symKeyId, nil
+		}
+	}
+	symKeyId, err := c.pss.GenerateSymmetricKey(notifier.topic, &address, false)
+	if err != nil {
+		return "", fmt.Errorf("Generate symkey fail: %v", err)
+	}
+	notifier.bins = append(notifier.bins, &sendBin{
+		address:  address,
+		symKeyId: symKeyId,
+		count:    1,
+	})
+	return symKeyId, nil
+}
+
+// Handler is the pss topic handler to be used to process notification service messages
+// It should be registered in the pss of both to any notification service provides and clients using the service
+func (c *Controller) Handler(smsg []byte, p *p2p.Peer, asymmetric bool, keyid string) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	log.Debug("notify controller handler", "keyid", keyid)
+
+	// see if the message is valid
+	msg := &Msg{}
+	err := rlp.DecodeBytes(smsg, msg)
+	if err != nil {
+		return fmt.Errorf("Invalid message: %v", err)
+	}
+
+	switch msg.Code {
+	case MsgCodeStart:
+		pubkey := crypto.ToECDSAPub(common.FromHex(keyid))
+
+		// if name is not registered for notifications we will not react
+		if _, ok := c.notifiers[msg.GetName()]; !ok {
+			return fmt.Errorf("Subscribe attempted on unknown resource %s", msg.GetName())
+		}
+
+		// parse the address from the message and truncate if longer than our mux threshold
+		address := msg.Payload
+		if len(msg.Payload) > c.notifiers[msg.GetName()].threshold {
+			address = address[:c.notifiers[msg.GetName()].threshold]
+		}
+
+		// add the address to the notification list
+		symKeyId, err := c.addToNotifier(msg.GetName(), address)
+		if err != nil {
+			return fmt.Errorf("add address to notifier fail: %v", err)
+		}
+		symkey, err := c.pss.GetSymmetricKey(symKeyId)
+		if err != nil {
+			return fmt.Errorf("retrieve symkey fail: %v", err)
+		}
+
+		// add to address book for send initial notify
+		pssaddr := pss.PssAddress(address)
+		err = c.pss.SetPeerPublicKey(pubkey, controlTopic, &pssaddr)
+		if err != nil {
+			return fmt.Errorf("add pss peer for reply fail: %v", err)
+		}
+
+		// send initial notify, will contain symkey to use for consecutive messages
+		notify, err := c.notifiers[msg.GetName()].contentFunc(msg.GetName())
+		if err != nil {
+			return fmt.Errorf("retrieve current update from source fail: %v", err)
+		}
+		replyMsg := NewMsg(MsgCodeNotifyWithKey, msg.GetName(), make([]byte, len(notify)+symKeyLength))
```

Comment:
why not just construct the payload and call the msg contructor with that.

Reply:
I don't understand what you mean